### PR TITLE
Minor improvements #2

### PR DESCRIPTION
**Pull request by mjauvin**
- improve error handling / messaging
- improve sso data getter/setter to support multiple attributes
- normalize sso email

### CODE REVIEW

**Comment by mjauvin** [comment]
@LukeTowers other incrmental improvements

File: controllers/Handle.php [hunk 1]
```diff
@@ -150,21 +168,53 @@ public function callback(string $provider): RedirectResponse
...
+    {
+        [$user, $domain] = explode('@', $email);
+
+        if (in_array($domain, ['gmail.com', 'googlemail.com'])) {
+            // Google emails can have "." anywhere in the username but the actual account has none.
+            $user = str_replace('.', '', $user);
+        }
+        # user+specifier@domain
+        # remove "+specifier" for all email accounts.
+        $user = preg_replace('#\+.+#', '', $user);
```
Comment (LukeTowers) [inline controllers/Handle.php:208]: This seems dangerous, why is this required?

File: controllers/Handle.php [hunk 2]
```diff
@@ -150,21 +168,53 @@ public function callback(string $provider): RedirectResponse
...
+
+        return $user . '@' . $domain;
+    }
+
+    public function redirectToSigninPage($msg = null): RedirectResponse
+    {
+        if ($msg) {
+            Flash::error($msg);
+        }
+        return Redirect::to(Session::pull('signin_url', Backend::url('backend/auth/signin')));
```
Comment (LukeTowers) [inline controllers/Handle.php:218]:
```suggestion
    public function redirectToSigninPage(string $message = null): RedirectResponse
    {
        if ($message) {
            Flash::error($message);
        }
        return Redirect::to(Session::pull('signin_url', Backend::url('backend/auth/signin')));
```

File: controllers/Handle.php (hunk 1 above)
Comment (mjauvin) [reply controllers/Handle.php:208]: @LukeTowers Why "dangerous" ?

In an email address, you can have anything after the + e.g. luke+test1@wintercms.com, luke+test2@wintercms.com, etc and they all deliver emails to luke@wintercms.com in the end. So it's better to remove those "extension" in order to have a "normalized" email address to work with.

Comment (LukeTowers) [reply controllers/Handle.php:208]: @mjauvin "Plus Addressing" is a feature that was specific to gmail, only recently (2020 or 2022?) added to MS 365; and is not part of any standards. https://mozilla.fandom.com/wiki/User:Me_at_work/plushaters

It's not safe to assume that just because there's a plus in the address "Plus Addressing" is being used and also you can have as many plus characters as you want in the local part.

The email is what the email is; why are we trying to "normalize" it beyond lowercasing it? (note that technically according to standards the local part could still be interpreted by mail servers as case sensitive if they wanted to, but realistically you're fine to treat the whole email as case insensitive: https://stackoverflow.com/a/9808332).

Comment (mjauvin) [reply controllers/Handle.php:208]: Might not be a standard, but I've used it for 25 years with postfix mail server and it's pretty well reckognised all over.

I can drop this if you prefer, or we could have an option that would do a more strict normalization.

Comment (LukeTowers) [reply controllers/Handle.php]: I guess at this point I'm not understanding what the value of "normalizing" this is, we should probably be fairly strict on the email matching logic when it comes to authentication lest we open ourselves up to potential attacks.

Comment (mjauvin) [reply controllers/Handle.php]: ok, I'm leaving the plus addressing alone.

File: controllers/Handle.php [hunk 3]
```diff
@@ -57,15 +58,26 @@ public function __construct()
     public function callback(string $provider): RedirectResponse
     {
         if (!in_array($provider, $this->enabledProviders)) {
-            abort(404);
+            return $this->redirectToSigninPage('This provider is not enabled');
```
Comment (LukeTowers) [inline controllers/Handle.php:61]:
```suggestion
            return $this->redirectToSigninPage("The {$provider} SSO provider is not enabled");
```

File: controllers/Handle.php [hunk 4]
```diff
@@ -57,15 +58,26 @@ public function __construct()
...
+        if (!Request::input('code')) {
+            $msg = 'Error: no access token was returned by provider (' . $provider . ')';
+            if ($error = Request::input('error')) {
+                $msg = $provider . ' error: ' . $error;
+                if ($errorDescription = Request::input('error_description')) {
+                    $msg .= ' (' . $errorDescription . ')';
+                }
+            }
+            return $this->redirectToSigninPage($msg);
+        }
```
Comment (LukeTowers) [inline controllers/Handle.php:78]:
```suggestion
        if (!Request::input('code')) {
            $message = 'Error: no access token was returned by provider (' . $provider . ')';
            if ($error = Request::input('error')) {
                $message = $provider . ' error: ' . $error;
                if ($errorDescription = Request::input('error_description')) {
                    $message .= ' (' . $errorDescription . ')';
                }
            }
            return $this->redirectToSigninPage($message);
        }
```

File: controllers/Handle.php [hunk 5]
```diff
@@ -57,15 +58,26 @@ public function __construct()
...
+        if (!Request::input('code')) {
+            $message = 'Error: no access token was returned by provider (' . $provider . ')';
+            if ($error = Request::input('error')) {
+                $message = $provider . ' error: ' . $error;
+                if ($errorDescription = Request::input('error_description')) {
+                    $message .= ' (' . $errorDescription . ')';
+                }
+            }
+            return $this->redirectToSigninPage($message);
+        }
```
Comment (LukeTowers) [inline controllers/Handle.php:78]: Where do the `error` and `error_description` values come from? Is that a standard feature of Socialite or is it specific to a particular provider?

File: controllers/Handle.php [hunk 6]
```diff
@@ -90,27 +103,32 @@ public function callback(string $provider): RedirectResponse
             // ]);
             // $user->setSsoConfig('allow_password_auth', false);
             // @TODO: Event here for registering user if desired, default fallback abort behaviour
-            abort(403, 'User not found');
+            $msg = 'Winter AuthManager: User not found (' . $email . ')';
+            return $this->redirectToSigninPage($msg);
```
Comment (LukeTowers) [inline controllers/Handle.php:107]:
```suggestion
            return $this->redirectToSigninPage("An account for $email could not be found.");
```

File: controllers/Handle.php (hunk 5 above)
Comment (mjauvin) [reply controllers/Handle.php:78]: I think this is in oAuth protocol, all providers seem to return those. I also found this:

https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/

File: controllers/Handle.php [hunk 7]
```diff
@@ -90,27 +103,31 @@ public function callback(string $provider): RedirectResponse
...
             // @TODO: Event here for registering user if desired, default fallback abort behaviour
-            abort(403, 'User not found');
+            return $this->redirectToSigninPage("An account for $email could not be found.");
         }
 
-        if (
-            $ssoUser->getId()
-            && $user->getSsoId($provider) !== $ssoUser->getId()
-        ) {
+        if ($ssoUser->getId() && $user->getSsoValue($provider, 'id') !== $ssoUser->getId()) {
```
Comment (LukeTowers) [inline controllers/Handle.php:109]:
```suggestion
        if (
            $ssoUser->getId()
            && $user->getSsoValue($provider, 'id') !== $ssoUser->getId()
        ) {
```

File: controllers/Handle.php [hunk 8]
```diff
@@ -90,27 +103,34 @@ public function callback(string $provider): RedirectResponse
...
             $user->save();
         }
 
         // Check if the user is allowed to keep a persistent session
         if (is_null($remember = Config::get('cms.backendForceRemember', false))) {
-            $remember = false;
-            // @TODO: Get this from the request
-            // $remember = (bool) post('remember');
+            // @TODO: needs to be saved to the Session on the signin form using an ajax request
+            $remember = Session::pull('backend.forceRemember', false);
```
Comment (LukeTowers) [inline controllers/Handle.php:121]: Is this implemented?

File: controllers/Handle.php [hunk 9]
```diff
@@ -57,15 +58,26 @@ public function __construct()
...
         // @TODO: Login or register the user / provide an event for plugins to handle
         // user registration themselves. Would like plugin to be able to handle frontend
         // or backend or even both. If event is used follow naming conventions from in progress
         // issues
 
-        $ssoUser = Socialite::driver($provider)->user();
+        if (!Request::input('code')) {
+            $message = 'Error: no access token was returned by provider (' . $provider . ')';
+            if ($error = Request::input('error')) {
+                $message = $provider . ' error: ' . $error;
```
Comment (LukeTowers) [inline controllers/Handle.php:72]:
```suggestion
                $message = $provider . ' SSO error: ' . $error;
```

File: Plugin.php [hunk 10]
```diff
@@ -94,13 +94,18 @@ protected function forceEmailLogin(): void
...
         User::extend(function ($model) {
-            $model->addDynamicMethod('getSsoId', function (string $provider) use ($model) {
-                return $model->metadata['winter.sso'][$provider]['id'] ?? null;
+            $model->addDynamicMethod('getSsoValue', function (string $provider, mixed $key, $default = null) use ($model) {
+                return $model->metadata['winter.sso'][$provider][$key] ?? $default;
             });
-            $model->addDynamicMethod('setSsoId', function (string $provider, string $id) use ($model) {
-                $metadata = $model->metadata ?? [];
-                $metadata['winter.sso'][$provider]['id'] = $id;
+            $model->addDynamicMethod('setSsoValues', function (string $provider, array $values, bool $save = false) use ($model) {
```
Comment (LukeTowers) [inline Plugin.php:100]: Is the save parameter used right now?

Comment (mjauvin) [reply Plugin.php:100]: No, it's going to be used in the next PR... ;)

File: controllers/Handle.php [hunk 11]
```diff
@@ -150,21 +170,48 @@ public function callback(string $provider): RedirectResponse
...
+        } catch (Exception $e) {
+            return $this->redirectToSigninPage($e->getMessage());
+        }
+        return $response;
+    }
+
+    /*
+     * Returns canonical form for google emails.
+     */
+    public function normalizeEmail($email)
```
Comment (LukeTowers) [inline controllers/Handle.php:199]:
```suggestion
    /**
     * Canonicalize the provided email based on domain name.
     */
    protected function normalizeEmail($email)
```

File: controllers/Handle.php (hunk 8 above)
Comment (mjauvin) [reply controllers/Handle.php:121]: No

File: controllers/Handle.php [hunk 12]
```diff
@@ -150,21 +170,48 @@ public function callback(string $provider): RedirectResponse
...
+        [$user, $domain] = explode('@', strtolower($email));
+
+        if (in_array($domain, ['gmail.com', 'googlemail.com'])) {
+            // Google emails can have "." anywhere in the username but the actual account has none.
+            $user = str_replace('.', '', $user);
+        }
+        return $user . '@' . $domain;
+    }
+
+    public function redirectToSigninPage(string $message = null): RedirectResponse
```
Comment (LukeTowers) [inline controllers/Handle.php:210]:
```suggestion
    protected function redirectToSigninPage(string $message = null): RedirectResponse
```

File: Plugin.php (hunk 10 above)
Comment (LukeTowers) [reply Plugin.php:100]: Can you add it in the next PR instead then?

Comment (LukeTowers) [reply Plugin.php:100]: I didn't catch that on the session signin_url usage and I would have preferred to have dealt with that in this PR when it gets used - makes for a cleaner history and easier to review.

File: controllers/Handle.php (hunk 8 above)
Comment (LukeTowers) [reply controllers/Handle.php:121]: Let's not have it here then. Functionally the difference between this and the comment is nil except that having this here without the other half of the implementation required is a bit of a code smell to me.

File: Plugin.php (hunk 10 above)
Comment (mjauvin) [reply Plugin.php]: done

File: controllers/Handle.php (hunk 8 above)
Comment (mjauvin) [reply controllers/Handle.php]: gone

File: controllers/Handle.php (hunk 1 above)
Comment (bennothommo) [reply controllers/Handle.php]: @mjauvin for what it's worth, I did a search up on whether plus aliases for emails have any ramifications for SSO purposes, and as far as SSO providers are concerned, there's no security issue with using them. Ultimately, it's up to the SSO provider whether they accept it or not (or handle it differently) - we shouldn't be stripping or otherwise modifying the email address when sending it to an SSO provider for authorisation.

Comment (mjauvin) [reply controllers/Handle.php]: that was done after receiving the email address from the SSO provider, before trying to match it against our db

Comment (bennothommo) [reply controllers/Handle.php]: Ah gotcha - I will just close me mouth then. :stuck_out_tongue:

File: controllers/Handle.php [hunk 13]
```diff
@@ -90,27 +104,31 @@ public function callback(string $provider): RedirectResponse
...
+            $user->setSsoValues($provider, ['id' => $ssoUser->getId()]);
             $user->save();
         }
 
         // Check if the user is allowed to keep a persistent session
-        if (is_null($remember = Config::get('cms.backendForceRemember', false))) {
-            $remember = false;
-            // @TODO: Get this from the request
-            // $remember = (bool) post('remember');
+        $remember = Config::get('cms.backendForceRemember', false);
```
Comment (LukeTowers) [inline controllers/Handle.php:120]:
```suggestion
        // @TODO: Support "null" as an option (where the user selects the remember me checkbox before logging in,
        // will probably require storing a flag in the session before redirecting them to the SSO provider login URL
        $remember = Config::get('cms.backendForceRemember', false);
```

**Comment by mjauvin** [comment]
@LukeTowers Any more concerns before merging this?

**Review by LukeTowers — APPROVED** [review]
LGTM!
